### ISSUE TITLE
Is it deadlock safe ?

### ISSUE THREAD

**Issue by Twodio** [outmatic/Dapper.UnitOfWork#5]
Hi again, last week i've forgot to ask you if there were deadlocks or any other async/sync tests for your async methods within sync methods.

In your [example](src/Dapper.UnitOfWork/Dapper.UnitOfWork.Example/Program.cs), line 36, you're calling the method MainAsync within a sync method, but i also read that the _console's Main method may be an exception to these cases where you can actually call an async method and get fine with it in most cases without causing the application to crash_.

`MainAsync(args).GetAwaiter().GetResult();`

The fact that this code will run synchronously may be a problem in most of the cases, after some reading on "don't block asynchronous code". As it looks, it may be there just for demonstration purposes and no real application at all, but i couldn't avoid to look at the same call in [Retry.cs](src/Dapper.UnitOfWork/Dapper.UnitOfWork/Retry.cs), line 34.

`HandleExceptionAsync(retryOptions, ex, retryCount).GetAwaiter().GetResult();`

After some reading and research about this method, i've found some warnings like _"they're meant to be used by the compiler"_ and _"they may cause deadlock"_.

All these approaches seems more like a way to unwrap the exceptions being thrown if any. My big question is: **Is it safe to use _GetAwaiter().GetResult()_ in this context ?**

**Comment by petrhaus** [comment]
Good catch. There are a few corner cases on which it could cause deadlocks, although I'm running it in a few high concurrency backend services without any (apparent) issue.
Anyway better avoid it and run in the threadpool via Task.Run() and then wait for the execution.
I'll try to update it asap (or feel free to open a pull if you have time).
Thanks.

**Comment by Twodio** [comment]
You're such a great person, always answering in record time. I already handled this issue using _Task.Run()_ in my examples, but as i already said i didn't have the chance to test your code even a bit, or mine fully. My code is now very different from yours since i only picked the main concept of yours(repository base) and from there everything changed. If you don't get the chance to update it today or during this week, i'll make the effort to send a pull request during these days or in the weekend if possible. I also noticed that you were asking for more examples, if i get the chance, i'll open a pull request for that aswell. Anyways your code looks geat man and helped me with my intern struggle on finding a better way to deal with such a concept, thanks for sharing that.
